### PR TITLE
applications: Asset Tracker v2: Add debug module that uses Memfault

### PR DESCRIPTION
**Pull request by simensrostad**
Add new debug module that subscribes to events from all modules in the
system. The module integrates basic support for Memfault with
custom GPS metrics in addition to the pre-existing metrics already
supported in NCS.

Closes CIA-355 and CIA-356

### CODE REVIEW

File: applications/asset_tracker_v2/src/events/Kconfig [hunk 1]
```diff
@@ -42,6 +42,10 @@ config DATA_EVENTS_LOG
 	bool "Enable logging for data module events"
 	default y
 
+config DEBUG_EVENTS_LOG
+	bool "Enable logging for Cloud module events"
```
Comment (balaji-nordic) [inline applications/asset_tracker_v2/src/events/Kconfig:46]: Cloud->Debug

File: applications/asset_tracker_v2/src/events/debug_module_event.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	DEBUG_EVT_MEMFAULT_DATA_READY,
+	DEBUG_EVT_ERROR
+};
+
+struct debug_module_memfault_data {
+	uint8_t buf[CONFIG_DEBUG_MODULE_MEMFAULT_CHUNK_SIZE_MAX];
+	size_t len;
+};
+
+/** @brief Debug event. */
```
Comment (balaji-nordic) [inline applications/asset_tracker_v2/src/events/debug_module_event.h:36]: I dont think this short @brief is needed.

File: applications/asset_tracker_v2/src/modules/Kconfig.cloud_module [hunk 3]
```diff
@@ -12,8 +12,8 @@ if CLOUD_MODULE
 
 config CLOUD_THREAD_STACK_SIZE
 	int "Cloud module thread stack size"
-	default 4096 if NRF_CLOUD
-	default 2688
+	default 6144 if NRF_CLOUD_MQTT
```
Comment (balaji-nordic) [inline applications/asset_tracker_v2/src/modules/Kconfig.cloud_module:15]: Just out of curiousity: What did we change NRF_CLOUD to NRF_CLOUD_MQTT in this PR? And why did the default stack size increase? If this is only because of debug module, then shouldn't we increase the thread size only if the debug module is configured?

File: applications/asset_tracker_v2/src/modules/Kconfig.debug_module [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+menuconfig DEBUG_MODULE
+	bool "Debug module"
+
+if DEBUG_MODULE
+
+config DEBUG_MODULE_USE_CUSTOM_TRANSPORT
+	bool "Use custom transport when sending memfault data"
```
Comment (balaji-nordic) [inline applications/asset_tracker_v2/src/modules/Kconfig.debug_module:13]: Should we not have a "default" value here?

File: applications/asset_tracker_v2/src/modules/Kconfig.debug_module [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	help
+	  If this option is disabled, Memfault firwmare SDKs own internal HTTP transport is used to
```
Comment (balaji-nordic) [inline applications/asset_tracker_v2/src/modules/Kconfig.debug_module:15]:
```suggestion
         If enabled, the debug module will forward the data in chunks via
	  events of type DEBUG_EVT_MEMFAULT_DATA_READY via the
	  configured cloud to Memfault. If disabled, the debug module will use the 
	  Memfault firmware SDKs internal HTTP transport. This can be used to save overhead 
	  associated with maintaining multiple connections.```

Comment (balaji-nordic) [inline applications/asset_tracker_v2/src/modules/Kconfig.debug_module:15]: Propose to word it this way.

          If enabled, the debug module will forward the data in chunks using
	  the events of type DEBUG_EVT_MEMFAULT_DATA_READY via the
	  configured cloud to Memfault. If disabled, the debug module will use the 
	  Memfault firmware SDKs internal HTTP transport. This can be used to save overhead 
	  associated with maintaining multiple connections.```

File: applications/asset_tracker_v2/src/modules/Kconfig.debug_module [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+	  multiple connections.
+
+config DEBUG_MODULE_MEMFAULT_HEARTBEAT_INTERVAL_SEC
+	int "Memfault heartbeat interval in seconds"
+	default 1800
+	help
+	  Configuration that sets how often Memfault captures and prepares data that will be sent
+	  to Memfault cloud. Reduce this interval to get more frequent updates to Memfault.
+	  Note that data transmitted to Memfault cloud is still dependent on how often the
+	  application sends data, which can be manipulated by setting the applications real-time
```
Comment (balaji-nordic) [inline applications/asset_tracker_v2/src/modules/Kconfig.debug_module:28]:
```suggestion
	  application sends data, which can be manipulated by setting the application's real-time
```

Comment (balaji-nordic) [reply applications/asset_tracker_v2/src/modules/Kconfig.debug_module:28]: What are those 'real-time configurations'?

File: applications/asset_tracker_v2/src/modules/Kconfig.debug_module (hunk 5 above)
Comment (balaji-nordic) [reply applications/asset_tracker_v2/src/modules/Kconfig.debug_module:15]: Also would'nt it always transmit "chunks" irrespective of the transport? If thats true, then you can just drop the wording about 'chunks'.

File: applications/asset_tracker_v2/src/modules/Kconfig.debug_module [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+	  configurations.
+
+endif
+
+config DEBUG_MODULE_MEMFAULT_CHUNK_SIZE_MAX
+	int "Fragment size of data transmitted to memfault"
+	default 2048 if DEBUG_MODULE_USE_CUSTOM_TRANSPORT
```
Comment (balaji-nordic) [inline applications/asset_tracker_v2/src/modules/Kconfig.debug_module:35]: Shoult it be so high when using custom transport?

File: applications/asset_tracker_v2/src/modules/Kconfig.debug_module [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+	default 80
+	help
+	  Decreasing this option will in case of a core-dump produce more events carrying
```
Comment (balaji-nordic) [inline applications/asset_tracker_v2/src/modules/Kconfig.debug_module:38]:
```suggestion
	  Decreasing this option will, in case of a core-dump, produce more events carrying
```

File: applications/asset_tracker_v2/src/modules/cloud_module.c [hunk 9]
```diff
@@ -392,6 +401,19 @@ static void data_send(struct data_module_event *evt)
 	send_data_ack(evt->data.buffer.buf, evt->data.buffer.len, true);
 }
 
+static void memfault_data_send(struct debug_module_event *evt)
+{
+	int err;
+
+	err = cloud_wrap_memfault_data_send(evt->data.memfault.buf, evt->data.memfault.len);
+	if (err) {
+		LOG_ERR("cloud_wrap_data_send, err: %d", err);
```
Comment (balaji-nordic) [inline applications/asset_tracker_v2/src/modules/cloud_module.c:410]:
```suggestion
		LOG_ERR("cloud_wrap_memfault_data_send, err: %d", err);
```

File: applications/asset_tracker_v2/src/modules/debug_module.c [hunk 10]
```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr.h>
+#include <logging/log.h>
+#if defined(CONFIG_MEMFAULT)
```
Comment (balaji-nordic) [inline applications/asset_tracker_v2/src/modules/debug_module.c:9]: defined(CONFIG_MEMFAULT) -> !defined(CONFIG_DEBUG_MODULE_USE_CUSTOM_TRANSPORT) ?

I am assuimg !DEBUG_MODULE_USE_CUSTOM_TRANSPORT implies CONFIG_MEMFAULT

File: applications/asset_tracker_v2/src/events/debug_module_event.h (hunk 2 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/events/debug_module_event.h:36]: Maybe not, I includeded it to align with the other module event docs in the application.

File: applications/asset_tracker_v2/src/modules/debug_module.c [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+		}
+	}
+
+#if defined(CONFIG_MEMFAULT)
+	memfault_handle_event(msg);
+#endif
+}
+
+EVENT_LISTENER(MODULE, event_handler);
+EVENT_SUBSCRIBE_EARLY(MODULE, app_module_event);
```
Comment (balaji-nordic) [inline applications/asset_tracker_v2/src/modules/debug_module.c:281]: Why does the debug module have to be an early subscriber to all the events? I was assuming that this be the late subscriber because the data processed and generated by this module is not delay sensitive.

File: applications/asset_tracker_v2/src/modules/gps_module.c [hunk 12]
```diff
@@ -52,6 +52,15 @@ static enum sub_state_type {
...
 static const struct device *gps_dev;
 
+static struct module_helpers {
+	/* Uptime set when GPS search is started. Used to calculate search time for
+	 * GPS fix and timeout.
+	 */
+	uint64_t start_uptime;
+	/* Numer of satellites tracked for a GPS fix or timeout. */
+	uint8_t satellites_tracked;
+} helpers;
```
Comment (balaji-nordic) [inline applications/asset_tracker_v2/src/modules/gps_module.c:62]: Why the name "helpers"? Shouldn't it rather be "stats" or "statistics"?

File: applications/asset_tracker_v2/src/modules/gps_module.c [hunk 13]
```diff
@@ -52,6 +52,15 @@ static enum sub_state_type {
 /* GPS device. Used to identify the GPS driver in the sensor API. */
 static const struct device *gps_dev;
 
+static struct module_helpers {
+	/* Uptime set when GPS search is started. Used to calculate search time for
+	 * GPS fix and timeout.
+	 */
+	uint64_t start_uptime;
+	/* Numer of satellites tracked for a GPS fix or timeout. */
```
Comment (balaji-nordic) [inline applications/asset_tracker_v2/src/modules/gps_module.c:60]:
```suggestion
	/* Number of satellites tracked before getting a GPS fix or a timeout. */
```

File: applications/asset_tracker_v2/src/modules/Kconfig.cloud_module (hunk 3 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/Kconfig.cloud_module:15]: After https://github.com/nrfconnect/sdk-nrf/pull/4609. Kconfigs associated with `NRF_CLOUD`  were split into separate module Kconfigs. `CLOUD_THREAD_STACK_SIZE` is now ultimately dependent on `NRF_CLOUD_MQTT` and not `NRF_CLOUD`. `NRF_CLOUD` is now more of a umbrella config for everything related to nRF Cloud libraries. We only care about the MQTT part in this case.

You`re right! The stack increase should depend on `CONFIG_DEBUG_MODULE_USE_CUSTOM_TRANSPORT` being enabled.

File: applications/asset_tracker_v2/src/modules/Kconfig.debug_module (hunk 4 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/Kconfig.debug_module:13]: Kconfig options automatically defaults to "n". So adding `default n` is not needed.

File: applications/asset_tracker_v2/src/modules/Kconfig.debug_module (hunk 5 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/Kconfig.debug_module:15]: Will update the description and remove the working around `chunks`.

File: applications/asset_tracker_v2/src/modules/Kconfig.debug_module (hunk 6 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/Kconfig.debug_module:28]: The applications real-time configurations are the configurations that governs the overall behavior of the device. They are documented in the root README.rst

File: applications/asset_tracker_v2/src/modules/Kconfig.debug_module (hunk 7 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/Kconfig.debug_module:35]: I'm afraid it has to. Or else the cloud module will get spammed with tens of events of type `DEBUG_EVT_MEMFAULT_DATA_READY` in case of a core-dump, carrying smaller amounts of data. Increasing this option makes the debug module send less events carrying more data. This way the internal processing queue in the cloud module won't get overfilled. (if that happens we must to a reboot to avoid undefined behavior in the app),.

File: applications/asset_tracker_v2/src/modules/debug_module.c (hunk 10 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/debug_module.c:9]: Actually `CONFIG_DEBUG_MODULE_USE_CUSTOM_TRANSPORT` depends on `CONFIG_MEMFAULT`. `CONFIG_DEBUG_MODULE_USE_CUSTOM_TRANSPORT` selects what transport Memfault data will be sent on. Ill update the Kconfig name to make this more clear

Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/debug_module.c:9]: Ill update it to `DEBUG_MODULE_MEMFAULT_USE_EXTERNAL_TRANSPORT` if no one disagrees.

File: applications/asset_tracker_v2/src/modules/debug_module.c [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+		if (err) {
+			LOG_ERR("memfault_software_watchdog_feed, error: %d", err);
+		}
+		break;
+	case WATCHDOG_EVT_TIMEOUT_INSTALLED:
+		/* Set the software watchdog timeout slightly shorter than the actual watchdog
+		 * timeout. This is to catch a timeout in advance so that Memfault can save
+		 * coredump data before a reboot occurs.
+		 */
+		err = memfault_software_watchdog_update_timeout(evt->timeout - 5000);
```
[the remark this answers is not in the document]
Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/debug_module.c:159]: Good idea. Ill add an assert.

File: applications/asset_tracker_v2/src/modules/debug_module.c (hunk 11 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/debug_module.c:281]: In some cases, modules consume events. Meaning that they are not carried around to other modules after the respective module that consumed the event has processed it. Subscribing early is just a precaution to ensure that the debug module most likely receives all events before they are potentially discarded.

File: applications/asset_tracker_v2/src/modules/gps_module.c (hunk 12 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/gps_module.c]: No opinion here really. I can rename it to `stats`

**Comment by simensrostad** [comment]
@umapraseeda Could you have a look at the documentation?

File: applications/asset_tracker_v2/src/modules/debug_module.c (hunk 10 above)
Comment (balaji-nordic) [reply applications/asset_tracker_v2/src/modules/debug_module.c:9]: Seems a bit wordy. But I cant think of any other better alternative. So go for it.

**Review by balaji-nordic — CHANGES_REQUESTED** [review]

File: applications/asset_tracker_v2/src/modules/gps_module.c [hunk 15]
```diff
@@ -52,6 +52,15 @@ static enum sub_state_type {
 /* GPS device. Used to identify the GPS driver in the sensor API. */
 static const struct device *gps_dev;
 
+static struct module_stats {
+	/* Uptime set when GPS search is started. Used to calculate search time for
+	 * GPS fix and timeout.
+	 */
+	uint64_t start_uptime;
+	/* Number of satellites tracked before getting a GPS fix or a timeout. */
```
[the remark this answers is not in the document]
Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/gps_module.c:60]: It should have been changed to your suggestion?

File: applications/asset_tracker_v2/src/modules/debug_module.c [hunk 16]
```diff
@@ -0,0 +1,299 @@
...
+#include "events/debug_module_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DEBUG_MODULE_LOG_LEVEL);
+
+#if defined(CONFIG_MEMFAULT)
+/* Time before the actual watchdog timeout that the Memfault software watchdog timeout will trigger.
+ * In milliseconds.
+ */
+#define MEMFAULT_SOFTWARE_WATCHDOG_DELTA 5000
```
[the remark this answers is not in the document]
Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/debug_module.c:38]: Jepp. Ill add a Kconfig

File: applications/asset_tracker_v2/src/modules/Kconfig.debug_module [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+	  If enabled, the debug module will forward captured Memfault data via events of type
+	  DEBUG_EVT_MEMFAULT_DATA_READY. The data should be forwarded via the
+	  configured cloud to Memfault. E.g via AWS IoT, Azure IoT Hub, or nRF Cloud.
+	  If this option is disabled, the debug module will use Memfault firwmare SDKs own internal
+	  HTTP transport. Transporting Memfault data over an already pre-established
+	  transport can save data associated with maintaining multiple connections.
+
+config DEBUG_MODULE_MEMFAULT_HEARTBEAT_INTERVAL_SEC
+	int "Memfault heartbeat interval in seconds"
+	default 1800
```
[the remark this answers is not in the document]
Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/Kconfig.debug_module:24]: Ill go through the Kconfigs and update potential dependencies. I'm also having some issue with the Unity test when depending on MEMFAULT.

File: applications/asset_tracker_v2/src/modules/gps_module.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (balaji-nordic) [reply applications/asset_tracker_v2/src/modules/gps_module.c:60]: You are right. This is fixed. For some reason, I did not see it.

File: applications/asset_tracker_v2/src/modules/Kconfig.debug_module [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+	help
+	  Configuration that sets how often Memfault captures and prepares data that will be sent
+	  to Memfault cloud. Reduce this interval to get more frequent updates to Memfault.
+	  Note that data transmitted to Memfault cloud is still dependent on how often the
+	  application sends data, which can be manipulated by setting the application's real-time
+	  configurations.
+
+endif
```
Comment (jtguggedal) [inline applications/asset_tracker_v2/src/modules/Kconfig.debug_module:32]: This `endif` should probably be at the end of the file

File: applications/asset_tracker_v2/src/modules/Kconfig.debug_module [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+	int "Fragment size of data transmitted to memfault"
+	default 2048 if DEBUG_MODULE_MEMFAULT_USE_EXTERNAL_TRANSPORT
+	default 80
+	help
+	  Decreasing this value, will in case of a core-dump, produce more events carrying
+	  Memfault data to the cloud module. This increases the possibility of an overfilled
+	  cloud module message queue depending on how fast the cloud module thread can handle
+	  incoming queue messages. An overfilled message queue is in general considered a fatal
+	  error that can cause undefined behaviour. To reduce the likelyhood of this issue
+	  occurring, this value should be kept reasonably high.
```
Comment (jtguggedal) [inline applications/asset_tracker_v2/src/modules/Kconfig.debug_module:44]: `, but within the modem firmware's documented limitations`

File: applications/asset_tracker_v2/src/modules/cloud_module.c [hunk 20]
```diff
@@ -91,7 +93,7 @@ const k_tid_t cloud_module_thread;
 static struct gps_agps_request agps_request;
 
 /* Cloud module message queue. */
-#define CLOUD_QUEUE_ENTRY_COUNT		10
+#define CLOUD_QUEUE_ENTRY_COUNT		20
```
Comment (jtguggedal) [inline applications/asset_tracker_v2/src/modules/cloud_module.c:96]: We should Kconfigify this configuration

File: applications/asset_tracker_v2/src/modules/debug_module.c [hunk 21]
```diff
@@ -0,0 +1,299 @@
...
+	}
+}
+#endif /* defined(CONFIG_WATCHDOG_APPLICATION) */
+
+static void send_memfault_data(void)
+{
+	bool data_available = memfault_packetizer_data_available();
+
+	if (IS_ENABLED(CONFIG_DEBUG_MODULE_MEMFAULT_USE_EXTERNAL_TRANSPORT) && data_available) {
+
```
Comment (jtguggedal) [inline applications/asset_tracker_v2/src/modules/debug_module.c:186]: redundant blank line

File: applications/asset_tracker_v2/tests/debug_module/src/debug_module_test.c [hunk 22]
```diff
@@ -0,0 +1,236 @@
...
+	TEST_ASSERT_EQUAL(0, DEBUG_MODULE_EVT_HANDLER((struct event_header *)app_module_event));
+
+	k_free(app_module_event);
+}
+
+void test_memfault_trigger_metric_sampling_on_gps_fix(void)
+{
+	setup_debug_module_in_init_state();
+
+	/* Update this function to expect the search time and number of satellites. */
```
Comment (balaji-nordic) [inline applications/asset_tracker_v2/tests/debug_module/src/debug_module_test.c:116]: Unwanted comment?

File: applications/asset_tracker_v2/tests/debug_module/src/debug_module_test.c [hunk 23]
```diff
@@ -0,0 +1,236 @@
...
+	__wrap_memfault_metrics_heartbeat_set_unsigned_ExpectAndReturn(
+						MEMFAULT_METRICS_KEY(GpsTimeoutSearchTime),
+						30000,
+						0);
+	__wrap_memfault_metrics_heartbeat_set_unsigned_ExpectAndReturn(
+						MEMFAULT_METRICS_KEY(GpsSatellitesTracked),
+						2,
+						0);
+	__wrap_memfault_metrics_heartbeat_debug_trigger_Ignore();
```
Comment (balaji-nordic) [inline applications/asset_tracker_v2/tests/debug_module/src/debug_module_test.c:152]: Shouldn't this be an Expect than an Ignore?

File: applications/asset_tracker_v2/tests/debug_module/src/debug_module_test.c [hunk 24]
```diff
@@ -0,0 +1,236 @@
...
+	__wrap_watchdog_register_handler_ExpectAnyArgs();
+	__wrap_watchdog_register_handler_AddCallback(&latch_watchdog_callback);
+	__wrap_module_start_ExpectAndReturn(&expected_module_data, 0);
+
+	TEST_ASSERT_EQUAL(0, DEBUG_MODULE_EVT_HANDLER((struct event_header *)app_module_event));
+
+	k_free(app_module_event);
+}
+
+void test_memfault_trigger_metric_sampling_on_gps_fix(void)
```
Comment (balaji-nordic) [inline applications/asset_tracker_v2/tests/debug_module/src/debug_module_test.c:112]: Can you write one or two lines about what each test case does?

File: applications/asset_tracker_v2/overlay-debug.conf [hunk 25]
```diff
@@ -64,3 +64,19 @@ CONFIG_NRF9160_GPS_LOG_LEVEL_DBG=y
...
 CONFIG_STREAM_FLASH_LOG_LEVEL_DBG=y
+
+# Debug module
+CONFIG_DEBUG_MODULE=y
+
+## Memfault - Used by the debug module
+# CONFIG_MEMFAULT=y
+# CONFIG_MEMFAULT_NCS_PROJECT_KEY=""
+# CONFIG_MEMFAULT_NCS_LTE_METRICS=y
+# CONFIG_MEMFAULT_NCS_STACK_METRICS=y
```
[the remark this answers is not in the document]
Comment (simensrostad) [reply applications/asset_tracker_v2/overlay-debug.conf:72]: Agreed, lets to that.

File: applications/asset_tracker_v2/tests/debug_module/src/debug_module_test.c [hunk 26]
```diff
@@ -0,0 +1,236 @@
...
+void test_memfault_unhandled_event(void)
+{
+	resetTest();
+	setup_debug_module_in_init_state();
+
+	/* Expect no memfault APIs to be called on GPS_EVT_ACTIVE */
+
+	struct gps_module_event *gps_module_event = new_gps_module_event();
+
+	gps_module_event->type = GPS_EVT_ACTIVE;
```
Comment (balaji-nordic) [inline applications/asset_tracker_v2/tests/debug_module/src/debug_module_test.c:197]: Can you extend this test to send all unhandled events? (If thats cumbersome, please change the name of this test to test_memfault_no_memfault_call_on_gps_evt_active())

**Review by balaji-nordic — CHANGES_REQUESTED** [review]

File: applications/asset_tracker_v2/src/cloud/aws_iot_integration.c [hunk 27]
```diff
@@ -32,6 +32,8 @@ LOG_MODULE_REGISTER(MODULE, CONFIG_CLOUD_INTEGRATION_LOG_LEVEL);
 #define PGPS_REQUEST_TOPIC_LEN (AWS_CLOUD_CLIENT_ID_LEN + 9)
 #define PGPS_RESPONSE_TOPIC "%s/pgps"
 #define PGPS_RESPONSE_TOPIC_LEN (AWS_CLOUD_CLIENT_ID_LEN + 5)
+#define MEMFAULT_TOPIC "%s/memfault"
```
[the remark this answers is not in the document]
Comment (simensrostad) [reply applications/asset_tracker_v2/src/cloud/aws_iot_integration.c:35]: Yes. Lets discuss this further when the cloud side is implemented.

File: applications/asset_tracker_v2/src/modules/cloud_module.c (hunk 20 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/cloud_module.c:98]: Agreed. Its in the backlog to do this for all modules in the application.

File: applications/asset_tracker_v2/tests/debug_module/src/debug_module_test.c (hunk 23 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/tests/debug_module/src/debug_module_test.c:152]: Jepp.

File: applications/asset_tracker_v2/tests/debug_module/src/debug_module_test.c (hunk 26 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/tests/debug_module/src/debug_module_test.c:197]: Jepp. will add test for all other unhandled events.

File: applications/asset_tracker_v2/prj.conf [hunk 28]
```diff
@@ -15,7 +15,7 @@ CONFIG_FPU=y
 # Heap and stacks
 CONFIG_HEAP_MEM_POOL_SIZE=32768
 CONFIG_MAIN_STACK_SIZE=1280
-CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=2048
+CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=3584
```
[the remark this answers is not in the document]
Comment (simensrostad) [reply applications/asset_tracker_v2/prj.conf:18]: its only necessary to increase this option if `DEBUG_MODULE_MEMFAULT_USE_EXTERNAL_TRANSPORT` if enabled. Moved the config to `overlay-memfault.conf`

File: applications/asset_tracker_v2/configuration/memfault/memfault_metrics_heartbeat_config.def [hunk 29]
```diff
@@ -0,0 +1,7 @@
+/* Application-specific metrics can be defined here.
+ * Please refer to https://docs.memfault.com/docs/embedded/metrics-api for more details.
```
[the remark this answers is not in the document]
Comment (simensrostad) [reply applications/asset_tracker_v2/configuration/memfault/memfault_metrics_heartbeat_config.def:2]: I like that it contains the links to the Memfault docs. But I don't have any strong opinion on this. I can remove it if you think its unnecessary.

File: applications/asset_tracker_v2/src/modules/Kconfig.debug_module (hunk 18 above)
Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/Kconfig.debug_module]: I removed the if guard for these statements as `DEBUG_MODULE_MEMFAULT_CHUNK_SIZE_MAX` is used in `debug_module_event.h` which is included in the `cloud_module`. Either we define a fixed size buffer in `debug_module_memfault_data` (which must be rather big) or we need to have `DEBUG_MODULE_MEMFAULT_CHUNK_SIZE_MAX` always exposed.

**Comment by simensrostad** [comment]
@jtguggedal @balaji-nordic _I think_ I've addressed all your review comments. @umapraseeda The documentation added in this PR should be ready for review.

File: applications/asset_tracker_v2/configuration/memfault/memfault_metrics_heartbeat_config.def (hunk 29 above)
[the remark this answers is not in the document]
Comment (jtguggedal) [reply applications/asset_tracker_v2/configuration/memfault/memfault_metrics_heartbeat_config.def:2]: Yeah, it has some value to have the link. It was mostly the ` can be defined here` part I found a bit misleading as there's already some metrics here. Maybe that can be cut, so that it reads `Application-specific metrics`, but it's not a bug deal. Don't push and re-run CI just for this change

File: applications/asset_tracker_v2/src/modules/Kconfig.debug_module [hunk 30]
```diff
@@ -0,0 +1,57 @@
+#
+# Copyright (c) 2021 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+menuconfig DEBUG_MODULE
+	bool "Debug module"
+
```
Comment (jtguggedal) [inline applications/asset_tracker_v2/src/modules/Kconfig.debug_module:9]: I think we need a guard here somewhere, otherwise all these options will show up in menuconfig and other UIs also when the debug module is disabled

**Review by balaji-nordic — APPROVED** [review]
Looks good now. 👍

**Review by jtguggedal — APPROVED** [review]
LGTM, thanks!

**Review by umapraseeda — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
